benchmarks/inference/llm-bench/src/llm_bench/benchmark_runner.py

```diff
@@ -0,0 +1,308 @@
+                }
+            )
+
+        all_config_product = []
+        for config in configs_list:
+            # Ensure all config values are iterable types (i.e., list or tuple)
+            for k, v in config.items():
+                if not isinstance(v, list) or isinstance(v, tuple):
+                    config[k] = [v]
+
+            # We treat num_clients differently to enable early stopping
+            num_clients = config.pop("num_clients")
+
+            # Generate all possible combinations of prompt config values
+            for vals in itertools.product(*[config[k] for k in prompt_config_keys]):
+                config_product = {k: v for k, v in zip(prompt_config_keys, vals)}
+                config_product["num_clients"] = num_clients
+                all_config_product.append(config_product)
+
+        logger.info(f"Generated {len(all_config_product)} benchmark run setting(s)")
+
+        for config in all_config_product:
+            num_clients = config.pop("num_clients")
+            prompt_config = PromptConfig(**config)
+            yield num_clients, prompt_config
+
+    def _generate_requests(self, prompt_config: PromptConfig, num_clients: int) -> None:
+        logger.info("Generating Prompts")
+
+        warmup_prompts = self.config.warmup_requests * num_clients
+        workload_prompts = self.config.num_requests_per_client * num_clients
+        for prompt in self.prompt_generator(
+            config=prompt_config, num_prompts=warmup_prompts + workload_prompts
+        ):
+            self.client_launcher.add_request(prompt)
+
+        logger.info(
+            f"Generated {warmup_prompts} warmup and {workload_prompts} workload prompts."
+        )
+
+    def _get_output_dir(self) -> Path:
+        return self.config.result_dir / self.config.api / self.config.model
+
+    def _get_output_path(self, prompt_config: PromptConfig, num_clients: int) -> Path:
+        output_dir = self._get_output_dir()
+        output_file = f"prompt{prompt_config.prompt_length}_gen{prompt_config.max_new_tokens}_clients{num_clients}.json"
+        return output_dir / output_file
+
+    def _save_results(self, prompt_config: PromptConfig, num_clients: int) -> None:
+        output_path = self._get_output_path(
+            prompt_config=prompt_config, num_clients=num_clients
+        )
+
+        logger.info(f"Saving results to {output_path}")
+
+        all_responses = []
+        while True:
+            try:
+                response = self.client_launcher.get_response()
+                all_responses.append(response.to_dict())
+            except queue.Empty:
+                break
+
+        os.makedirs(output_path.parent, exist_ok=True)
+        with open(output_path, "w") as fh:
+            json.dump(all_responses, fh, indent=2)
+
+        logger.info(f"Saved {len(all_responses)} responses to {output_path}")
+
+    def run(self) -> None:
+        # Start the client service
```

Comment:
It seems we no longer check if the results exist or not (aka skip the test configs that seem already benchmarked). I think that feature would be very helpful when resuming an interrupted run.

Reply:
Excellent point, I will add that feature back!